### PR TITLE
Jetty 9.4.x 6254 queued request total timeout

### LINKED ISSUE #6254: Total timeout not enforced for queued requests
**Jetty version**
9.4.x

**Description**
The total timeout for requests that are queued in a `HttpDestination` is not always enforced.
Sending a request just before a previous queued one is about to expire resets the timeout of the already queued one.

### PR DESCRIPTION
**Pull request by sbordet**
@gregw first commit is cleanup and javadocs improvements, second commit is the actual fix.

### CODE REVIEW

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java [hunk 1]
```diff
@@ -564,14 +564,17 @@ private TimeoutTask(Scheduler scheduler)
...
 
-            // Check all queued exchanges for those that have expired
-            // and to determine when the next check must be.
+            // Reset the earliest timeout so we can expire again.
+            // A concurrent call to schedule(long) may lose an earliest
+            // value, but the corresponding exchange is already enqueued
+            // and will be seen by scanning the exchange queue below.
+            earliestTimeout.set(Long.MAX_VALUE);
+
+            long earliest = Long.MAX_VALUE;
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]:
```suggestion
            long earliest = Long.MAX_VALUE;
            earliestTimeout.set(earliest);
```

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java [hunk 2]
```diff
@@ -580,34 +583,27 @@ public void onTimeoutExpired()
...
-            // If subsequently that exchange is removed from the queue, the
-            // timeout is not cancelled, instead the entire queue is swept
-            // for expired exchanges and a new timeout is set.
-            long timeoutAt = nextTimeout.getAndUpdate(e -> Math.min(e, expiresAt));
-            if (timeoutAt != expiresAt)
+            // Schedule a timeout for the earliest exchange that may expire.
+            // When the timeout expires, scan the exchange queue for the next
+            // earliest exchange that may expire, and reschedule a new timeout.
+            long earliest = earliestTimeout.getAndUpdate(t -> Math.min(t, expiresAt));
+            if (expiresAt < earliest)
```
Comment (gregw) [inline jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: Is this right?   If expiresAt is after earliestTimeout, then earliest == earliestTimeout, which is < expiresAt, so do nothing... correct.
But if expiresAt is before earliestTimeout, then earliest == expiresAt, so the < also fails and we also do nothing - wrong?

Surely we need to do a schedule IFF earliest==expiresAt ?

```suggestion
            if (expiresAt == earliest)
```

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java [hunk 3]
```diff
@@ -102,8 +119,9 @@ public boolean schedule(long delay, TimeUnit units)
             if (_timeout.compareAndSet(timeout, new Timeout(newTimeoutAt, wakeup)))
             {
                 if (LOG.isDebugEnabled())
-                    LOG.debug("Installed timeout in {} ms, waking up in {} ms",
+                    LOG.debug("Installed timeout in {} ms, {} wake up in {} ms",
```
Comment (gregw) [inline jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java:123]: Use {} on multi-line log

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java (hunk 2 above)
Comment (sbordet) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: @gregw the update is a get-then-update, so `earliest` is the _old_ minimum. We don't know the new minimum, could be either `earliest` or `expireAt`.

The less-than check determines what the new minimum is; if it's `expireAt` then we need to schedule.

File: jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java (hunk 3 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java:123]: I am using `{}`, not sure what you mean here?

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java (hunk 2 above)
Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: Ah, then `earliest` is the wrong name.  Should be `lastEarliest`

File: jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java (hunk 3 above)
Comment (gregw) [reply jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java:123]:
```java
if (LOG.isDebugEnabled())
{
   LOG.debug(something,
   somethingelse?or:something
   blah,
   blah);
}
```

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java (hunk 2 above)
Comment (gregw) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: or `expiringAt`

**Review by gregw — DISMISSED** [review]
Other than the niggles in my comments, this is good.... but please change the variable name before merging

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java (hunk 1 above)
Comment (sbordet) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: Done.

File: jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java (hunk 2 above)
Comment (sbordet) [reply jetty-client/src/main/java/org/eclipse/jetty/client/HttpDestination.java]: `prevEarliest`

File: jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java (hunk 3 above)
Comment (sbordet) [reply jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeout.java:123]: Done.

**Pull request merged**